File: app/src/main/java/com/projet/sluca/smallbrother/PhoneStatReceiver.java
package com.projet.sluca.smallbrother;

import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.telephony.TelephonyManager;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

public class PhoneStatReceiver extends BroadcastReceiver
{
    static String callNumber = "";
    static UserData userdata = new UserData();   // Liaison avec les données globales de l'utilisateur.

    @Override
    public void onReceive(Context context, Intent intent)
    {
        userdata.loadData();

        if(userdata.getTelephone() != null)
        {
            if(intent.getAction().equals(Intent.ACTION_NEW_OUTGOING_CALL))
            {
            /*
            incomingFlag = false;
            String phoneNumber = intent.getStringExtra(Intent.EXTRA_PHONE_NUMBER);
            Log.i(TAG, "call OUT:"+phoneNumber);
            */
            }
            else
            {
                TelephonyManager tm =(TelephonyManager)context.getSystemService(Service.TELEPHONY_SERVICE);

                switch (tm.getCallState())
                {
                    case TelephonyManager.CALL_STATE_RINGING:

                        // Set du numéro de l'appelant.
                        this.callNumber = intent.getStringExtra("incoming_number");

                        if(SmsReceiver.getBit() == 1) // Si le Mode Privé est activé.
                        {
                            // Avertir :
                            SmsReceiver.setBit(3); // cas d'un appel
                        }
                        else
                        {
                            // Déclaration d'un passage dans la WorkActivity pour éviter que, au retour dans
                            // AideActivity, ne soit généré un doublon du Handler local.
                            userdata.setEsquive(true);

                            // lancement de la "WorkActivity".
                            Intent intnt = new Intent(context, WorkActivity.class);
                            intnt.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                            context.startActivity(intnt);
                        }
                }
            }
        }
    }

    static String catchcallNumber() { return callNumber; }  // Getter du numéro de l'appelant.
    static void resetCallNumber()   { callNumber = ""; }    // Réinitialiser.
}

File: app/src/main/java/com/projet/sluca/smallbrother/PhoneStatReceiver.kt
package com.projet.sluca.smallbrother

import android.app.Application
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.telephony.TelephonyManager
import com.projet.sluca.smallbrother.activities.WorkActivity
import com.projet.sluca.smallbrother.models.UserData

/**
 * PhoneStatReceiver manages information when a phone call is received on the user's phone
 *
 * @author Sébastien Luca & Maxime Caucheteur (Updated on 08-01-2023)
 */
class PhoneStatReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        //userData = UserDataManager.getUserData(context.applicationContext as Application)
        userData.loadData(context)
        if (intent.action != Intent.ACTION_NEW_OUTGOING_CALL) processCall(context, intent)
    }

    /**
     * Process the phone call
     * @param [context] the context of the activity
     * @param [intent] the intent on receive of a call
     * @author Maxime Caucheteur
     * @version 1.2 (Updated on 04-01-2023)
     */
    private fun processCall(context: Context, intent: Intent) {
        val tm = context.getSystemService(Service.TELEPHONY_SERVICE) as TelephonyManager
        if (tm.callState == TelephonyManager.CALL_STATE_RINGING) {
            if(intent.hasExtra("incoming_number")) callNumber = intent
                .getStringExtra("incoming_number").toString()
            if (userData.bit == 1) userData.bit = 3 // Private mode ON
            else {
                userData.esquive = true
                val intnt = Intent(context, WorkActivity::class.java)
                intnt.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                context.startActivity(intnt)
            }
        }
    }

    companion object {
        var callNumber = ""
        lateinit var userData: UserData 

        fun catchCallNumber(): String = callNumber
        fun resetCallNumber() {
            callNumber = ""
        }
    }
}
